app/models/assignment.rb

```diff
@@ -1247,29 +1247,122 @@ def to_json(options = {})
+  # +zip_file+. Also writes the starter files for this assignment in the same directory as +settings_filepath+.
+  def starter_file_config_to_zip(zip_file, settings_filepath)
+    default_starter_group = nil
+    group_data = []
+    directory_path = File.dirname(settings_filepath)
+    self.starter_file_groups.each.with_index(1) do |starter_file_group, download_count|
+      group_name = "#{download_count}: #{starter_file_group.name}"
+      group_name = group_name.gsub(File::SEPARATOR, '')
+      starter_file_group.write_starter_files_to_zip(zip_file, File.join(directory_path, group_name))
+      if starter_file_group.id == self.default_starter_file_group_id
+        default_starter_group = group_name
+      end
+      group_data << {
+        directory_name: group_name,
+        name: starter_file_group.name,
+        use_rename: starter_file_group.use_rename,
+        entry_rename: starter_file_group.entry_rename
+      }
+    end
+    starter_file_settings = {
+      starter_file_type: self.starter_file_type,
+      allow_starter_files_after_due: self.starter_files_after_due,
+      default_starter_group: default_starter_group,
+      group_information: group_data
+    }.to_yaml
+    zip_file.get_output_stream(settings_filepath) { |f| f.write starter_file_settings }
+  end
+
   # zip all files in the folder at +self.autotest_files_dir+ and return the
   # path to the zip file
   def zip_automated_test_files(user)
     zip_name = "#{self.short_identifier}-testfiles-#{user.user_name}"
     zip_path = File.join('tmp', zip_name + '.zip')
     FileUtils.rm_rf zip_path
-    files_dir = Pathname.new self.autotest_files_dir
     Zip::File.open(zip_path, create: true) do |zip_file|
-      self.autotest_files.map do |file|
-        path = File.join zip_name, file
-        abs_path = files_dir.join(file)
-        if abs_path.directory?
-          zip_file.mkdir(path)
-        else
-          zip_file.get_output_stream(path) { |f| f.print File.read(abs_path.to_s, mode: 'rb') }
-        end
-      end
+      add_test_files_to_zip(zip_file, zip_name)
     end
     zip_path
   end
 
+  # Writes all of this assignment's automated test files to the +zip_dir+ in +zip_file+. Also writes
+  # the tester settings specified in this assignment's properties and spec file to the json file at
+  # +settings_file_path+ in the +zip_file+.
+  def automated_test_config_to_zip(zip_file, zip_dir, settings_file_path)
+    self.add_test_files_to_zip(zip_file, zip_dir)
+    test_specs_path = self.autotest_settings_file
+    test_specs = File.exist?(test_specs_path) ? JSON.parse(File.open(test_specs_path, &:read)) : {}
+    test_specs['testers']&.each do |tester_info|
+      tester_info['test_data']&.each do |test_info|
+        test_info['extra_info']&.delete('test_group_id')
+      end
+    end
+    misc_settings = {}
+    misc_settings_attr = [:enable_test, :enable_student_tests, :tokens_per_period, :unlimited_tokens,
```

Comment:
For the same reason as in the comment for assignment_properties_config I would like to avoid explicitly including attribute names whenever possible. I know that you want to keep all the information about automated tests together in one zip file entry but in this case, these attributes should really be with the assignment properties, not here.

If we don't set them initially when we create the assignment then we run the risk of the new assignment not being valid (for example, if one of these attributes becomes a required attributes later on)